File: app/services/external/google_directions_service.py
"""Google Directions API (legacy REST) for road-following map polylines."""

from __future__ import annotations

from dataclasses import dataclass

import httpx
import structlog

from app.core.config import settings
from app.schemas.departure_plan import TransportMode

logger = structlog.get_logger(__name__)

_DIRECTIONS_URL = "https://maps.googleapis.com/maps/api/directions/json"


@dataclass(frozen=True)
class MapRouteOption:
    encoded_polyline: str
    distance_km: float
    travel_minutes: int
    provider: str
    route_index: int
    summary: str | None = None


@dataclass(frozen=True)
class MapDirectionsResult:
    encoded_polyline: str
    distance_km: float
    travel_minutes: int
    provider: str = "google_directions_backend"
    routes: tuple[MapRouteOption, ...] = ()


class GoogleDirectionsBackendService:
    """Server-side wrapper around Google Directions JSON API."""

    async def get_directions(
        self,
        origin_lat: float,
        origin_lng: float,
        dest_lat: float,
        dest_lng: float,
        mode: TransportMode = TransportMode.driving,
    ) -> MapDirectionsResult:
        api_key = settings.GOOGLE_MAPS_API_KEY or settings.GOOGLE_MAPS_SERVER_API_KEY
        if not api_key:
            raise RuntimeError("GOOGLE_MAPS_API_KEY is not configured")

        params = {
            "origin": f"{origin_lat},{origin_lng}",
            "destination": f"{dest_lat},{dest_lng}",
            "mode": _travel_mode(mode),
            "alternatives": "true",
            "key": api_key,
        }

        async with httpx.AsyncClient(timeout=8.0) as client:
            response = await client.get(_DIRECTIONS_URL, params=params)

        response.raise_for_status()
        payload = response.json()

        status = (payload.get("status") or "UNKNOWN").upper()
        if status != "OK":
            error_message = payload.get("error_message") or "Directions request failed"
            logger.warning(
                "google_directions_backend.failed",
                status=status,
                error=error_message,
            )
            raise RuntimeError(f"Google Directions API error: {status} — {error_message}")

        raw_routes = payload.get("routes") or []
        if not raw_routes:
            raise RuntimeError("Google Directions API returned no routes")

        parsed_routes: list[MapRouteOption] = []
        for index, route in enumerate(raw_routes):
            option = _parse_route(route, index)
            if option is not None:
                parsed_routes.append(option)

        if not parsed_routes:
            raise RuntimeError("Google Directions API returned no valid overview polylines")

        primary = parsed_routes[0]

        logger.info(
            "google_directions_backend.success",
            distance_km=primary.distance_km,
            travel_minutes=primary.travel_minutes,
            polyline_length=len(primary.encoded_polyline),
            mode=mode.value,
            route_count=len(parsed_routes),
        )

        return MapDirectionsResult(
            encoded_polyline=primary.encoded_polyline,
            distance_km=primary.distance_km,
            travel_minutes=primary.travel_minutes,
            provider="google_directions_backend",
            routes=tuple(parsed_routes),
        )


def _parse_route(route: dict, index: int) -> MapRouteOption | None:
    encoded = ((route.get("overview_polyline") or {}).get("points") or "").strip()
    if not encoded:
        return None

    legs = route.get("legs") or []
    distance_meters = 0
    duration_seconds = 0
    for leg in legs:
        distance_meters += int(((leg.get("distance") or {}).get("value")) or 0)
        duration_seconds += int(((leg.get("duration") or {}).get("value")) or 0)

    summary = (route.get("summary") or "").strip() or None

    return MapRouteOption(
        encoded_polyline=encoded,
        distance_km=round(distance_meters / 1000, 1),
        travel_minutes=max(1, round(duration_seconds / 60)),
        provider="google_directions_backend",
        route_index=index,
        summary=summary,
    )


def _travel_mode(mode: TransportMode) -> str:
    if mode == TransportMode.walking:
        return "walking"
    if mode == TransportMode.transit:
        return "transit"
    return "driving"

File: Flutter/app/lib/core/services/google_directions_service.dart
import 'package:dio/dio.dart';
import 'package:flutter/foundation.dart';
import 'package:flutter_polyline_points/flutter_polyline_points.dart';
import 'package:google_maps_flutter/google_maps_flutter.dart';

import '../constants/app_constants.dart';
import '../network/api_client.dart';

enum GoogleTravelMode { driving, walking, transit }

class GoogleDirectionsRoute {
  const GoogleDirectionsRoute({
    required this.points,
    required this.distanceMeters,
    required this.durationSeconds,
    this.encodedPolyline,
    this.provider,
    this.routeIndex = 0,
    this.summary,
  });

  final List<LatLng> points;
  final int distanceMeters;
  final int durationSeconds;
  final String? encodedPolyline;
  final String? provider;
  final int routeIndex;
  final String? summary;

  double get distanceKm => distanceMeters / 1000;

  int get durationMinutes =>
      durationSeconds <= 0 ? 0 : (durationSeconds / 60).ceil().clamp(1, 9999);
}

class GoogleDirectionsException implements Exception {
  GoogleDirectionsException(this.message, {this.status});

  final String message;
  final String? status;

  @override
  String toString() => status == null ? message : '$message ($status)';
}

/// Fetches road-following routes.
///
/// - Web: uses backend /maps/directions.
/// - Mobile/desktop native: tries backend first, then Google Directions REST.
///
/// Important:
/// This service rejects routes with only 2 points because those are usually
/// straight-line fallback routes, not real road-following routes.
class GoogleDirectionsService {
  GoogleDirectionsService({
    PolylinePoints? polylinePoints,
    Dio? dio,
    ApiClient? apiClient,
    String? apiKey,
    Duration timeout = const Duration(seconds: 12),
  }) : _polylinePoints = polylinePoints ?? PolylinePoints(),
       _dio = dio ?? Dio(),
       _apiClient = apiClient,
       _apiKey = apiKey ?? AppConstants.googleMapsApiKey,
       _timeout = timeout;

  final PolylinePoints _polylinePoints;
  final Dio _dio;
  final ApiClient? _apiClient;
  final String _apiKey;
  final Duration _timeout;

  bool get isConfigured => _apiKey.trim().isNotEmpty;

  List<LatLng> decodePolyline(String encoded) {
    if (encoded.trim().isEmpty) {
      return const [];
    }

    return _polylinePoints
        .decodePolyline(encoded)
        .map((point) => LatLng(point.latitude, point.longitude))
        .toList(growable: false);
  }

  /// This is only fallback.
  /// Do not treat it as a real road route because it returns only 2 points.
  List<LatLng> straightLineFallback(LatLng origin, LatLng destination) {
    return [origin, destination];
  }

  Future<GoogleDirectionsRoute?> tryGetRoute({
    required LatLng origin,
    required LatLng destination,
    GoogleTravelMode mode = GoogleTravelMode.driving,
  }) async {
    final alternatives = await tryGetRouteAlternatives(
      origin: origin,
      destination: destination,
      mode: mode,
    );

    if (alternatives.isEmpty) {
      return null;
    }

    return _fastestRoute(alternatives);
  }

  Future<List<GoogleDirectionsRoute>> tryGetRouteAlternatives({
    required LatLng origin,
    required LatLng destination,
    GoogleTravelMode mode = GoogleTravelMode.driving,
  }) async {
    final backendRoutes = await _tryBackendRouteAlternatives(
      origin: origin,
      destination: destination,
      mode: mode,
    );

    if (backendRoutes.isNotEmpty) {
      debugPrint('Backend route alternatives count: ${backendRoutes.length}');
      return backendRoutes;
    }

    if (!kIsWeb && isConfigured) {
      try {
        final directRoutes = await getRouteAlternatives(
          origin: origin,
          destination: destination,
          mode: mode,
        );

        debugPrint('Google direct route alternatives count: ${directRoutes.length}');
        return directRoutes;
      } on GoogleDirectionsException catch (e) {
        debugPrint('Google Directions exception: $e');
      } on Exception catch (e) {
        debugPrint('Google Directions unknown error: $e');
      }
    }

    return const [];
  }

  Future<List<GoogleDirectionsRoute>> _tryBackendRouteAlternatives({
    required LatLng origin,
    required LatLng destination,
    required GoogleTravelMode mode,
  }) async {
    final client = _apiClient;
    if (client == null) {
      return const [];
    }

    try {
      final response = await client
          .get<Map<String, dynamic>>(
            '/maps/directions',
            queryParameters: {
              'origin_lat': origin.latitude,
              'origin_lng': origin.longitude,
              'dest_lat': destination.latitude,
              'dest_lng': destination.longitude,
              'mode': _modeQuery(mode),
            },
          )
          .timeout(_timeout);

      final data = response.data ?? const {};
      final rawRoutes = data['routes'] as List<dynamic>? ?? const [];

      if (rawRoutes.isNotEmpty) {
        final parsed = <GoogleDirectionsRoute>[];
        for (final item in rawRoutes) {
          if (item is! Map<String, dynamic>) {
            continue;
          }
          final route = _parseBackendRouteOption(item);
          if (route != null) {
            parsed.add(route);
          }
        }
        if (parsed.isNotEmpty) {
          return parsed;
        }
      }

      final fallbackRoute = _parseBackendRouteOption(data);
      return fallbackRoute == null ? const [] : [fallbackRoute];
    } on Exception catch (e) {
      debugPrint('Backend directions error: $e');
      return const [];
    }
  }

  GoogleDirectionsRoute? _parseBackendRouteOption(Map<String, dynamic> data) {
    final encoded = data['encoded_polyline'] as String?;
    if (encoded == null || encoded.trim().isEmpty) {
      debugPrint('Backend route: encoded polyline is empty');
      return null;
    }

    final points = decodePolyline(encoded);

    debugPrint('Backend decoded points count: ${points.length}');
    debugPrint('Backend encoded polyline length: ${encoded.length}');

    if (points.length <= 2) {
      debugPrint('Backend route rejected because it has <= 2 points');
      return null;
    }

    final distanceKm = (data['distance_km'] as num?)?.toDouble() ?? 0;
    final travelMinutes = (data['travel_minutes'] as num?)?.round() ?? 0;

    return GoogleDirectionsRoute(
      points: points,
      distanceMeters: (distanceKm * 1000).round(),
      durationSeconds: travelMinutes * 60,
      encodedPolyline: encoded,
      provider: data['provider'] as String? ?? 'backend',
      routeIndex: (data['route_index'] as num?)?.round() ?? 0,
      summary: data['summary'] as String?,
    );
  }

  Future<GoogleDirectionsRoute> getRoute({
    required LatLng origin,
    required LatLng destination,
    GoogleTravelMode mode = GoogleTravelMode.driving,
  }) async {
    final routes = await getRouteAlternatives(
      origin: origin,
      destination: destination,
      mode: mode,
    );

    if (routes.isEmpty) {
      throw GoogleDirectionsException('Directions returned no valid routes');
    }

    return _fastestRoute(routes);
  }

  Future<List<GoogleDirectionsRoute>> getRouteAlternatives({
    required LatLng origin,
    required LatLng destination,
    GoogleTravelMode mode = GoogleTravelMode.driving,
  }) async {
    if (!isConfigured) {
      throw GoogleDirectionsException(
        'Google Maps API key is missing. '
        'Pass --dart-define=MAPS_API_KEY=YOUR_KEY',
      );
    }

    final response = await _dio
        .get<Map<String, dynamic>>(
          'https://maps.googleapis.com/maps/api/directions/json',
          queryParameters: {
            'origin': '${origin.latitude},${origin.longitude}',
            'destination': '${destination.latitude},${destination.longitude}',
            'mode': _modeQuery(mode),
            'alternatives': 'true',
            'key': _apiKey,
          },
          options: Options(responseType: ResponseType.json),
        )
        .timeout(_timeout);

    final payload = response.data ?? const {};
    final status = payload['status'] as String? ?? 'UNKNOWN';

    debugPrint('Google Directions status: $status');

    if (status.toUpperCase() != 'OK') {
      throw GoogleDirectionsException(
        payload['error_message'] as String? ?? 'Directions request failed',
        status: status,
      );
    }

    final routes = payload['routes'] as List<dynamic>? ?? const [];
    if (routes.isEmpty) {
      throw GoogleDirectionsException(
        'Directions returned no routes',
        status: status,
      );
    }

    final parsed = <GoogleDirectionsRoute>[];
    for (var index = 0; index < routes.length; index++) {
      final route = routes[index];
      if (route is! Map<String, dynamic>) {
        continue;
      }

      final parsedRoute = _parseGoogleRouteOption(route, index);
      if (parsedRoute != null) {
        parsed.add(parsedRoute);
      }
    }

    if (parsed.isEmpty) {
      throw GoogleDirectionsException('Directions returned no valid route points');
    }

    return parsed;
  }

  GoogleDirectionsRoute? _parseGoogleRouteOption(
    Map<String, dynamic> route,
    int routeIndex,
  ) {
    final encodedOverview =
        ((route['overview_polyline'] as Map<String, dynamic>?)?['points']
            as String?);

    final stepPoints = _decodeStepPolylines(route);

    final points = stepPoints.isNotEmpty
        ? stepPoints
        : (encodedOverview == null
              ? const <LatLng>[]
              : decodePolyline(encodedOverview));

    if (points.length <= 2) {
      debugPrint(
        'Google route $routeIndex rejected because it has <= 2 points',
      );
      return null;
    }

    final legs = route['legs'] as List<dynamic>? ?? const [];

    var distanceMeters = 0;
    var durationSeconds = 0;

    for (final leg in legs) {
      final legMap = leg as Map<String, dynamic>;

      distanceMeters += (legMap['distance'] as Map?)?['value'] as int? ?? 0;
      durationSeconds += (legMap['duration'] as Map?)?['value'] as int? ?? 0;
    }

    return GoogleDirectionsRoute(
      points: points,
      distanceMeters: distanceMeters,
      durationSeconds: durationSeconds,
      encodedPolyline: encodedOverview,
      provider: 'google_directions',
      routeIndex: routeIndex,
      summary: route['summary'] as String?,
    );
  }

  GoogleDirectionsRoute _fastestRoute(List<GoogleDirectionsRoute> routes) {
    final sorted = List<GoogleDirectionsRoute>.from(routes)
      ..sort((a, b) {
        final durationCompare = a.durationSeconds.compareTo(b.durationSeconds);
        if (durationCompare != 0) {
          return durationCompare;
        }
        return a.distanceMeters.compareTo(b.distanceMeters);
      });
    return sorted.first;
  }

  List<LatLng> _decodeStepPolylines(Map<String, dynamic> route) {
    final legs = route['legs'] as List<dynamic>? ?? const [];
    final merged = <LatLng>[];

    for (final leg in legs) {
      final legMap = leg as Map<String, dynamic>;

      final steps = legMap['steps'] as List<dynamic>? ?? const [];

      for (final step in steps) {
        final stepMap = step as Map<String, dynamic>;

        final encoded =
            ((stepMap['polyline'] as Map<String, dynamic>?)?['points']
                as String?) ??
            '';

        if (encoded.isEmpty) continue;

        final decoded = decodePolyline(encoded);

        if (decoded.isEmpty) continue;

        if (merged.isNotEmpty && _samePoint(merged.last, decoded.first)) {
          merged.addAll(decoded.skip(1));
        } else {
          merged.addAll(decoded);
        }
      }
    }

    return merged;
  }

  bool _samePoint(LatLng a, LatLng b) {
    return (a.latitude - b.latitude).abs() < 0.000001 &&
        (a.longitude - b.longitude).abs() < 0.000001;
  }

  String _modeQuery(GoogleTravelMode mode) {
    return switch (mode) {
      GoogleTravelMode.walking => 'walking',
      GoogleTravelMode.transit => 'transit',
      GoogleTravelMode.driving => 'driving',
    };
  }
}
